File: shoring/desk/src/main/java/com/panopset/marin/secure/checksums/ChecksumReport.java
package com.panopset.marin.secure.checksums;

import java.io.File;
import java.util.List;

import com.panopset.compat.ChecksumGenerator;
import com.panopset.compat.Fileop;
import com.panopset.compat.Logop;
import com.panopset.compat.Nls;
import com.panopset.compat.Padop;
import com.panopset.compat.Stringop;
import com.panopset.util.rpg.TextProcessor;

public class ChecksumReport {

  private boolean firstTime = true;

  public void generateReport(File file, List<ChecksumType> types, TextProcessor textProcessor) {
    if (file == null) {
      Logop.warn(Nls.xlate("File is null, no checksum report generated."));
      return;
    }
    if (!file.exists()) {
      Logop.warn(
          String.join(":", Nls.xlate("File doesn't exist"), Fileop.getCanonicalPath(file)));
      return;
    }
    if (file.isDirectory()) {
      generateDirectoryReport(file, types, textProcessor);
    } else {
      generateFileReport(file, types, textProcessor);
    }
  }

  private Integer maxTitleLength;

  private int getMaxTitleSize(List<ChecksumType> types) {
    if (maxTitleLength == null) {
      int i = 0;
      for (ChecksumType cst : types) {
        if (cst.name().length() > i) {
          i = cst.name().length();
        }
      }
      maxTitleLength = i;
    }
    return maxTitleLength;
  }

  private void generateFileReport(File file, List<ChecksumType> types,
      TextProcessor textProcessor) {
    Logop.green(String.join(": ", Nls.xlate("Processing"), Fileop.getCanonicalPath(file)));
    if (types.size() > 1) {
      textProcessor.append(file.getName());
      textProcessor.append("\n");
    } else if (types.size() == 1) {
      if (firstTime) {
        textProcessor.append(types.get(0).name());
        textProcessor.append("\n");
        firstTime = false;
      }
    }
    for (ChecksumType cst : types) {
      if (types.size() > 1) {
        textProcessor.append(Padop.leftPad(cst.name(), getMaxTitleSize(types), ' '));
        textProcessor.append(":");
      }
      switch (cst) {
        case BYTES:
          textProcessor.append(ChecksumGenerator.byteCount(file));
          break;
        case MD5:
          textProcessor.append(ChecksumGenerator.md5(file));
          break;
        case SHA1:
          textProcessor.append(ChecksumGenerator.sha1(file));
          break;
        case SHA256:
          textProcessor.append(ChecksumGenerator.sha256(file));
          break;
        case SHA384:
          textProcessor.append(ChecksumGenerator.sha384(file));
          break;
        case SHA512:
          textProcessor.append(ChecksumGenerator.sha512(file));
          break;
        default:
          break;
      }
      if (types.size() > 1) {
        textProcessor.append("\n");
      } else {
        textProcessor.append(" ");
        textProcessor.append(file.getName());
      }
    }
    Logop.green(String.format("%s: %s", Nls.xlate("Completed"), Fileop.getCanonicalPath(file)));
  }

  private void generateDirectoryReport(File file, List<ChecksumType> types,
      TextProcessor textProcessor) {
    for (File f : file.listFiles()) {
      if (f.isFile()) {
        generateFileReport(f, types, textProcessor);
        textProcessor.append(Stringop.getEol());
      }
    }
    if (textProcessor.getText().isEmpty()) {
      Logop.warn(
          String.format(Stringop.CS, Nls.xlate("No files found in"), Fileop.getCanonicalPath(file)));
    } else {
      Logop.green(Nls.xlate("Completed."));
    }
  }
}

File: code/shoring/desk/src/main/kotlin/com/panopset/marin/secure/checksums/ChecksumReport.kt
package com.panopset.marin.secure.checksums

import com.panopset.compat.*
import com.panopset.compat.Fileop.getCanonicalPath
import java.io.File

class ChecksumReport(private val textProcessor: TextProcessor) {
    private var firstTime = true
    fun generateReport(file: File, types: List<ChecksumType>) {
        textProcessor.clear()
        if (!file.exists()) {
            Logz.warn(
                java.lang.String.join(":", Nls.xlate("File doesn't exist"), getCanonicalPath(file))
            )
            return
        }
        if (file.isDirectory) {
            generateDirectoryReport(file, types)
        } else {
            generateFileReport(file, types)
        }
    }

    private var maxTitleLength: Int = 0
    private fun getMaxTitleSize(types: List<ChecksumType>): Int {
            var i = 0
            for (cst in types) {
                if (cst.name.length > i) {
                    i = cst.name.length
                }
            }
            maxTitleLength = i
        return maxTitleLength
    }

    private fun generateFileReport(
        file: File, types: List<ChecksumType>
    ) {
        Logz.green(java.lang.String.join(": ", Nls.xlate("Processing"), getCanonicalPath(file)))
        if (types.size > 1) {
            textProcessor.append(file.name)
            textProcessor.append("\n")
        } else if (types.size == 1) {
            if (firstTime) {
                textProcessor.append(types[0].name)
                textProcessor.append("\n")
                firstTime = false
            }
        }
        for (cst in types) {
            if (types.size > 1) {
                textProcessor.append(Padop.leftPad(cst.name, getMaxTitleSize(types), ' '))
                textProcessor.append(":")
            }
            when (cst) {
                ChecksumType.BYTES -> textProcessor.append(byteCount(file))
                ChecksumType.MD5 -> textProcessor.append(md5(file))
                ChecksumType.SHA1 -> textProcessor.append(sha1(file))
                ChecksumType.SHA256 -> textProcessor.append(sha256(file))
                ChecksumType.SHA384 -> textProcessor.append(sha384(file))
                ChecksumType.SHA512 -> textProcessor.append(sha512(file))
            }
            if (types.size > 1) {
                textProcessor.append("\n")
            } else {
                textProcessor.append(" ${file.name}\n")
            }
        }
        Logz.green(String.format("%s: %s", Nls.xlate("Completed"), getCanonicalPath(file)))
    }

    private fun generateDirectoryReport(
        file: File, types: List<ChecksumType>
    ) {
        var fount = false
        if (!file.isDirectory) {
            return
        }
        val fileList = file.listFiles() ?: return
        for (f in fileList) {
            if (f.isFile) {
                fount = true
                generateFileReport(f, types)
            }
        }
        if (fount) {
            Logz.green(Nls.xlate("Completed."))
        } else {
            Logz.warn(String.format(Stringop.CS, Nls.xlate("No files found in"), getCanonicalPath(file)))
        }
    }
}
